KAFKA-7660: fix parentSensors memory leak

vvcephei opened the pull request:
In StreamsMetricsImpl, the parentSensors map was keeping references to Sensors after the sensors themselves had been removed.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
KAFKA-7660: fix parentSensors memory leak

vvcephei commented:
Note: I still plan to write a unit test for this.

vvcephei committed:
unit test

vvcephei commented:
@guozhangwang @bbejeck do you mind taking a look at this if you have the chance?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -413,12 +413,17 @@ public void removeSensor(final Sensor sensor) {
         }
     }
 
+    /** visible for testing */
```
nit: `/** ... */` -> `/* */`

ijuma replied:
We usually avoid the get prefix in cases like this

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -413,12 +413,17 @@ public void removeSensor(final Sensor sensor) {
         metrics.removeSensor(sensor.name());
 
-        final Sensor parent = parentSensors.get(sensor);
+        final Sensor parent = parentSensors.remove(sensor);
```
nice catch!

mjsax reviewed (approved):
LGTM.

bbejeck reviewed (approved):
Thanks for the patch @vvcephei LGMT

bbejeck commented:
failures unrelated

retest this please

mjsax commented:
@vvcephei Build failed with checkstyle error.

vvcephei committed:
clean up and test creation and removal of x-Level sensors

vvcephei committed:
cr comments

vvcephei commented:
Hi all, I've addressed the comments and fixed the error.

I found an additional leak in `removeAllNodeLevelSensors` that left an empty `Deque` for each node that had been unregistered.

I took the opportunity to refactor the other (`xLevelSensor`) tracking mechanisms we use and added a test to make sure the metrics really do get cleaned up and idempotently registered during migrations. The refactored code is simpler and should be more resilient to mistakes like this.

I don't plan to tack on anything else to this PR.

Do you mind taking another look?

Thanks,
-John

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java:
```diff
@@ -62,6 +71,60 @@ public void testRemoveSensor() {
+
+        assertThat(registry.metrics().size(), equalTo(numberOfTaskMetrics));
+
+        final Sensor parent2 = metrics.taskLevelSensor(taskName, operation, Sensor.RecordingLevel.DEBUG);
```
What is this used for? Sensor `parent1` and `parent2` should be referring to exactly the same object right?

vvcephei replied:
They should be, but in reality the same parent sensor is (re)registered each time a child sensor is created, so I wanted to (idempotently) create a second parent sensor to make sure there's no extra references.

guozhangwang reviewed (commented):
@vvcephei Thanks for the updated patch! I just had a minor question on unit test and lgtm otherwise.

bbejeck reviewed (approved):
Changes LGTM

vvcephei commented:
The test failures are unrelated:
```
kafka.api.ClientIdQuotaTest.testThrottledProducerConsumer
kafka.api.DelegationTokenEndToEndAuthorizationTest.testNoConsumeWithDescribeAclViaAssign
kafka.api.UserQuotaTest.testThrottledProducerConsumer
kafka.server.DynamicBrokerReconfigurationTest.testAddRemoveSaslListeners
```

Retest this, please.

miguno commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java:
```diff
@@ -62,6 +71,60 @@ public void testRemoveSensor() {
+    }
+
+    @Test
+    public void testMutiLevelSensorRemoval() {
```
Typo: `testMu[l]tiLevelSensorRemoval() {`
